tests/std-checks/core/src/ptr.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Add contracts for functions from std::ptr.
+
+use std::ptr::NonNull;
+
+/// Create wrapper functions to standard library functions that contains their contract.
+pub mod contracts {
+    use super::*;
+    use kani::{ensures, implies, mem::*, modifies, requires};
+
+    #[ensures(implies!(ptr.is_null() => result.is_none()))]
+    #[ensures(implies!(!ptr.is_null() => result.is_some()))]
+    pub fn new<T>(ptr: *mut T) -> Option<NonNull<T>> {
+        NonNull::new(ptr)
+    }
+
+    /// # Safety:
+    /// When calling this method, you have to ensure that all the following is true:
+    ///
+    ///  - The pointer must be properly aligned.
+    ///  - It must be “dereferenceable” in the sense defined in the module documentation.
+    ///  - TODO: The pointer must point to an initialized instance of T.
+    ///
+    /// You must enforce Rust’s aliasing rules, since the returned lifetime 'a is arbitrarily chosen and does not
+    /// necessarily reflect the actual lifetime of the data. In particular, while this reference exists, the memory
+    /// the pointer points to must not get mutated (except inside UnsafeCell).
+    /// Taken from: <https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.as_ref>
+    #[requires(assert_valid_ptr(obj.as_ptr()))]
```

Comment:
I totally agree with that, and I don't like the name either.

> Anything used inside a requires or ensures clause should work for both assertions and assumptions. 

The reality is that we cannot implement in Kani today a predicate that states whether a pointer is valid or not in every context. Changing the name of this function won't change that.

I think the current name accurately represents what is implemented today, that this must be done to assert that a pointer is valid.

Reply:
The other possibility is to create an unsafe `is_valid_ptr` which requires users to ensures that the base of the pointer points to a valid allocation or it is null. I do find that even less elegant, but it would at least capture the semantics of the operation implemented today.

Reply:
How about having a special version of requires, e.g. `requires_valid_ptr` which Kani can expand to an assertion for stubbing and to allocation for verification?

Reply:
I've actually thought about this, and I'm OK to rename this function and assert that the pointer is valid for 0 byte access, and document that this is an existing limitation in Kani. Hopefully, we can unblock this PR.

I would be OK with an annotation `requires_valid_ptr`, or something similar. That said, I have some concerns with the memory allocation part. I have concerns about the lifetime of this potential allocation, as well as how this would handle aliasing.